src/main/drivers/flash_w25n01g.c

```diff
@@ -580,6 +592,174 @@ void w25n01g_pageProgramFinish(flashDevice_t *fdevice)
+        {.u.link = {NULL, NULL}, 0, true, NULL},
+    };
+
+    static busSegment_t segmentsDataLoad[] = {
+        {.u.buffers = {readStatus, readyStatus}, sizeof(readStatus), true, w25n01g_callbackReady},
+        {.u.buffers = {writeEnable, NULL}, sizeof(writeEnable), true, w25n01g_callbackWriteEnable},
+        {.u.buffers = {progExecDataLoad, NULL}, sizeof(progExecDataLoad), false, NULL},
+        {.u.buffers = {NULL, NULL}, 0, true, NULL}, // Patch in pointer to data buffer here
+        {.u.link = {NULL, NULL}, 0, true, NULL},
+    };
+
+    static busSegment_t segmentsRandomDataLoad[] = {
+        {.u.buffers = {readStatus, readyStatus}, sizeof(readStatus), true, w25n01g_callbackReady},
+        {.u.buffers = {writeEnable, NULL}, sizeof(writeEnable), true, w25n01g_callbackWriteEnable},
+        {.u.buffers = {progRandomProgDataLoad, NULL}, sizeof(progRandomProgDataLoad), false, NULL},
+        {.u.buffers = {NULL, NULL}, 0, true, NULL}, // Patch in pointer to data buffer here
+        {.u.link = {NULL, NULL}, 0, true, NULL},
+    };
+
+    busSegment_t *programSegment;
+
+    // Ensure any prior DMA has completed before continuing
+    spiWait(fdevice->io.handle.dev);
+
+    uint32_t columnAddress;
+
+    if (bufferDirty) {
+        columnAddress = W25N01G_LINEAR_TO_COLUMN(programLoadAddress);
+        // Set the address and buffer details for the random data load
+        progRandomProgDataLoad[1] = (columnAddress >> 8) & 0xff;
+        progRandomProgDataLoad[2] = columnAddress & 0xff;
+        segmentsRandomDataLoad[3].u.buffers.txData = (uint8_t *)buffers[0];
+        segmentsRandomDataLoad[3].len = bufferSizes[0];
+
+        programSegment = segmentsRandomDataLoad;
+    } else {
+        programStartAddress = programLoadAddress = fdevice->currentWriteAddress;
+        columnAddress = W25N01G_LINEAR_TO_COLUMN(programLoadAddress);
+        // Set the address and buffer details for the data load
+        progExecDataLoad[1] = (columnAddress >> 8) & 0xff;
+        progExecDataLoad[2] = columnAddress & 0xff;
+        segmentsDataLoad[3].u.buffers.txData = (uint8_t *)buffers[0];
+        segmentsDataLoad[3].len = bufferSizes[0];
+
+        programSegment = segmentsDataLoad;
+    }
+
+    bufferDirty = true;
+    programLoadAddress += bufferSizes[0];
+
+    if (W25N01G_LINEAR_TO_COLUMN(programLoadAddress) == 0) {
+        // Flash the loaded data
+        currentPage = W25N01G_LINEAR_TO_PAGE(programStartAddress);
+
+        progExecCmd[2] = (currentPage >> 8) & 0xff;
+        progExecCmd[3] = currentPage & 0xff;
+
+        // Don't callback on completion of data load but rather after flashing
+        programSegment[3].callback = NULL;
+
+        spiLinkSegments(fdevice->io.handle.dev, programSegment, segmentsFlash);
+
+        bufferDirty = false;
+
+        programStartAddress = programLoadAddress;
+    } else {
+        // Callback on completion of data load
+        programSegment[3].callback = w25n01g_callbackWriteComplete;
```

Comment:
I'd play it safe and clear link explicitly. BUS_ERROR (or future refactoring) may prevent link clearing code to work, and consequences can be ugly.
```
        spiLinkSegments(NULL, programSegment, NULL);
```

Reply:
The clearing is done automatically at https://github.com/betaflight/betaflight/blob/f4d6a2ce4344233284db50f89ef9389a1ac11166/src/main/drivers/bus_spi.c#L441

This is needed because the linking is the mechanism for queuing requests and such linkage between segments from different drivers MUST be broken like this. This can't ever be refactored away.

Reply:
(Caller may be required to clear last link of inserted chain. But as cycle prevention, ISR clearing is good idea.
It will get really interesting if one driver ends with BUS_ERROR - linked chains are not terminated and unlinked.)

IMO it is slightly better to break link explicitly here - normal links are created by driver chaining code, that code is responsible to undo it. But here the link was created explicitly, it is safer to undo it.